### PR TITLE
Implement tests for `SYCL_KHR_DEFAULT_CONTEXT` extension

### PR DESCRIPTION
**Pull request by 0x12CC**
Implement the tests from https://github.com/KhronosGroup/SYCL-CTS/issues/944.

### CODE REVIEW

File: tests/extension/khr_default_context/default_context.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace default_context::tests {
+
+#ifdef SYCL_KHR_DEFAULT_CONTEXT
+
+static void testDefaultContext() {
+  sycl::platform platform{};
+  sycl::context defaultContext = platform.khr_get_default_context();
+  CHECK(defaultContext.get_devices() == platform.get_devices());
+}
```
Comment (psalz) [inline tests/extension/khr_default_context/default_context.cpp:31]: If we decide that trying to compile a KHR extension test for a SYCL implementation that does not implement said extension should result in an error, I'd prefer the tests to use a more idiomatic style, e.g.:

```suggestion
TEST_CASE("The default context contains all of the platform's devices") {
  sycl::platform platform{};
  sycl::context defaultContext = platform.khr_get_default_context();
  CHECK(defaultContext.get_devices() == platform.get_devices());
}
```

File: tests/extension/khr_default_context/default_context.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        sycl::queue{syclContext, syclDevice, propList}.get_context());
+  CHECK(syclContext ==
+        sycl::queue{syclContext, syclDevice, asyncHandler, propList}
+            .get_context());
+}
+
+#endif
+
+TEST_CASE("Test case for \"Default Context\" extension",
+          "[khr_default_context") {
```
Comment (psalz) [inline tests/extension/khr_default_context/default_context.cpp:70]:
```suggestion
          "[khr_default_context]") {
```

**Review by psalz — COMMENTED** [review]
Since this is the first KHR extension test, we should think about how we want to approach this in general. Currently you have both a CMake option to enable/disable compilation of tests for the extension, as well as compile-time guards that check whether the extension is available. I'm wondering whether we need both.

If we keep the former (which I think is a good idea, at least for now -- we'll see whether it scales), I'd argue we can get rid of the `#ifdef SYCL_KHR_DEFAULT_CONTEXT`. If someone wants to test the extension, they'd likely want to know that it isn't available during compilation, not at runtime.

**Comment by gmlueck** [comment]
> If we keep the former (which I think is a good idea, at least for now -- we'll see whether it scales), I'd argue we can get rid of the `#ifdef SYCL_KHR_DEFAULT_CONTEXT`. If someone wants to test the extension, they'd likely want to know that it isn't available during compilation, not at runtime.

I think this makes sense.  If the CTS is run with `SYCL_CTS_ENABLE_EXT_KHR_DEFAULT_CONTEXT_TESTS`, then it seems like the test suite should fail if the implementation does not define `SYCL_KHR_DEFAULT_CONTEXT`.  Otherwise, an implementation could pass the CTS for this extension even without implementing it.

**Comment by psalz** [comment]
> If I get rid of this macro, someone that wants to test an implementation that implements most features and not this one would need to enumerate all of the relevant CMake options. They wouldn't be able to use the SYCL_CTS_ENABLE_KHR_TESTS option since this test would fail to compile.

Is that really a common use-case though? Realistically, I would assume that it's mostly going to be Intel that runs the CTS for the time being. If it does get out of hand, we could always provide [CMake presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html) for the supported implementations.

> Catch2 distinguishes passing test cases from skipped ones.

True, I'm just not sure whether having that skip message is worth sacrificing legibility of the code by either grouping the tests into guarded functions instead of using individual `TEST_CASE`s (like you did in this PR), or by littering `#ifdef`s all over the place.

**Comment by psalz** [comment]
Of course we could also get fancy and do a configuration-time check ([`try_compile`](https://cmake.org/cmake/help/v3.31/command/try_compile.html)) to see whether an extension is available, and if not, compile a proxy translation unit that only prints the skipped message.

**Comment by gmlueck** [comment]
> Catch2 distinguishes passing test cases from skipped ones. Here's the output for this test when the macro is not defined:
> 
> ```
> SYCL-CTS/tests/extension/khr_default_context/default_context.cpp:72: SKIPPED:
> explicitly with message:
>   SYCL_KHR_DEFAULT_CONTEXT is not defined
> 
> ================================================================================
> test cases: 1 | 1 skipped
> assertions: - none -
> ```

This requires a human to examine the test output, though.  I think it would be better for the test to fail (return a non-zero exit status) if the macro isn't defined.  This will make the tests easier to run in a CI system, which just tests the exit status.

I think it would also be good for the Khronos workgroup when a vendor submits a conformance report.  The WG won't want to look through the test output to see if there are SKIPPED messages.  It will be easier for the WG to just rely on the fact that the CTS passed when determining a vendor's conformance.

> If you both still agree that the SYCL_KHR_DEFAULT_CONTEXT macro guard is not needed then I'll remove it from the test.

I think we should go a step further.  We should remove the macro guard, and also add a test that explicitly tests that the macro is defined.  Part of an extension's API is that the macro is supposed to be set, so the CTS should check this.

> Also, if we do choose to remove it, should we remove the macros from the other extension tests (in a different PR)?

The other extensions are all "oneapi" extensions, right?  Putting my Khronos hat on ... these extensions exist in the Khronos CTS as a convenience to DPC++, so DPC++ can decide what works best for them.

Putting my DPC++ hat on ... I think this probably does make sense.  We've had situations in the past where we forgot to define the macro for an extension.  Restructuring the tests like this would catch these errors.  Of course, we'd need to have some cmake option to control when the oneapi tests are run.  Do we have something like `SYCL_CTS_ENABLE_ONEAPI_TESTS` already?

> If I get rid of this macro, someone that wants to test an implementation that implements most features and not this one would need to enumerate all of the relevant CMake options. They wouldn't be able to use the SYCL_CTS_ENABLE_KHR_TESTS option since this test would fail to compile.

One more comment about this.  When vendors submit official conformance results, I think we do *not* want them to use `SYCL_CTS_ENABLE_KHR_TESTS` because it's not obvious which extensions are being tested.  Wouldn't we want official runs to pass a list of `SYCL_CTS_ENABLE_KHR_DEFAULT_CONTEXT_TESTS`, ... options, so we know exactly which extensions the vendor is claiming conformance?

File: tests/extension/khr_default_context/default_context.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+#include "../../common/common.h"
+
+namespace default_context::tests {
+
+TEST_CASE(
+    "the default context extension defines the SYCL_KHR_DEFAULT_CONTEXT macro",
+    "[khr_default_context]") {
+#ifndef SYCL_KHR_DEFAULT_CONTEXT
+  static_assert(false, "SYCL_KHR_DEFAULT_CONTEXT is not defined");
```
Comment (psalz) [inline tests/extension/khr_default_context/default_context.cpp:29]: I think this is a good solution!

File: tests/extension/khr_default_context/default_context.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+          "[khr_default_context]") {
+  const sycl::property_list& propList = {};
+  cts_async_handler asyncHandler;
+  const auto& deviceSelector = sycl::default_selector_v;
+  sycl::device syclDevice;
+  sycl::context syclContext;
+  sycl::context defaultContext = sycl::platform{}.khr_get_default_context();
+
+  // Check that a default-constructed context is not the default context.
+  CHECK(syclContext != defaultContext);
```
Comment (psalz) [inline tests/extension/khr_default_context/default_context.cpp:50]: Minor: This could be pulled out into it's own test case IMO.

**Review by psalz — APPROVED** [review]

**Review by gmlueck — APPROVED** [review]
I think this looks great!

If you are on a roll with tests for default context behavior, we also need tests for #943.  Note that those tests are for the core spec, so they are not tied to the extension.

**Comment by tomdeakin** [comment]
Thanks! Waiting for further feedback from AdaptiveCpp.

**Review by illuhad — APPROVED** [review]

**Comment by tomdeakin** [comment]
WG approved to merge.

**Pull request merged**